### PR TITLE
Fix sharding strategy for distributed DL

### PR DESCRIPTION
**Pull request by ejguan**
1. Change the sharding strategy from sharding by worker first then by rank to sharding in the order of rank then workers.
2. Change to fetch Rank and World size in main process for the sake of `spawn`.

For the change 1:
Before this PR, for the case when dataset can not be evenly divided by `worker_num * world_size`, more data will be retrieved by workers in first RANKs.
Using the following example:
- dataset size: 100
- world_size: 4
- num_worker: 2

The number of data retrieved by each rank before this PR
- Rank 0: 26
- Rank 1: 26
- Rank 2: 24
- Rank 3: 24

The number of data retrieved by each rank after this PR
- Rank 0: 25
- Rank 1: 25
- Rank 2: 25
- Rank 3: 25

For the change 2:
Before this PR, `dist` functions are invoked inside worker processes. It's fine when the worker processes are forked from the parent process. All environment variables are inherited and exposed to these `dist` functions. However, when the worker processes are spawned, they won't be able to access to these environment variables, then the dataset won't be sharded by rank.
After this PR, `_sharding_worker_init_fn` should be working for both `spawn` and `fork` case.

### CODE REVIEW

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -92,25 +92,27 @@ def __iter__(self):
...
     info = torch.utils.data.get_worker_info()
     total_workers = info.num_workers
     datapipe = info.dataset
-    if dist.is_available() and dist.is_initialized():
-        total_workers *= dist.get_world_size()
-        global_worker_id = dist.get_rank() * info.num_workers + global_worker_id
+    # To distribute elements across distributed process evenly, we should shard data on distributed
+    # processes first then shard on worker processes
+    total_workers *= world_size
+    global_worker_id = global_worker_id * world_size + rank_id
```
Comment (ejguan) [inline torch/utils/data/dataloader.py:110]: Change sharding order here.

File: torch/utils/data/dataloader.py [hunk 2]
```diff
@@ -246,19 +248,19 @@ def __init__(self, dataset: Dataset[T_co], batch_size: Optional[int] = 1,
         # 2. Additional worker init function will take care of sharding in MP and Distributed
         if isinstance(self.dataset, IterDataPipe):
             self.dataset = _IterDataPipeSerializationWrapper(self.dataset)
+            ws, rank = _get_distributed_settings()
```
Comment (ejguan) [inline torch/utils/data/dataloader.py:251]: Change fetching distribute settings in main process here.

**Review by VitalyFedyunin — APPROVED** [review]

**Review by NivekT — APPROVED** [review]
LGTM!

I just walked through your example to confirm.

* dataset size: 100
* world_size: 4
* num_worker: 2

In `DataLoader`, we will have:
```python
ws = 4
rank =  0  # or 1, 2, 3
```

Then, in `_sharding_worker_init_fn`:
```python
world_size = 4
rank_id = 0  # or 1, 2, 3
worker_id = 0  # or 1
```
Then
```python
global_worker_id = 0  # or 1
total_workers = 2  # We have two workers
total_workers *= 4 = 8
global_worker_id = 0  # or 1, 2, ..., 7
```
Each rank two worker IDs should be (0, 4) (or (1, 5), etc). Then, when we apply sharding, there should be an even number of data per rank.

**Pull request closed without merging**